src/index/autotune.rs

```diff
@@ -0,0 +1,91 @@
+use std::ffi;
+use std::ptr;
+
+use super::*;
+
+use crate::error::Result;
+use crate::faiss_try;
+
+/// Uses a-priori knowledge on the Faiss indexes to extract tunable parameters.
+pub struct ParameterSpace {
+    inner: *mut FaissParameterSpace,
+}
+
+impl ParameterSpace {
+    /// Create a new instance
+    pub fn new() -> Result<Self> {
+        unsafe {
+            let mut inner = ptr::null_mut();
+            faiss_try(faiss_ParameterSpace_new(&mut inner))?;
+
+            Ok(ParameterSpace { inner })
+        }
+    }
+
+    /// Set one of the parameters
+    pub fn set_index_parameter(
+        &self,
+        index: &mut IndexImpl,
```

Comment:
This would constrain the method to index value of type `IndexImpl`. The only thing needed in the implementation is the inner pointer to an index, which is provided by the trait `NativeIndex`. So we can make the method more generic with something like this:

```rust
    pub fn set_index_parameter<I>(
        &self,
        index: I,
        ....
    )
where
    I: NativeIndex,
```
```